packages/curve-ui-kit/src/entities/campaigns/campaigns.ts

```diff
@@ -0,0 +1,118 @@
+import { useMemo } from 'react'
+import type { Address } from 'viem'
+import type { Chain } from '@curvefi/prices-api'
+import { fromEntries } from '@curvefi/prices-api/objects.util'
+import { useQueries } from '@tanstack/react-query'
+import { combineQueriesMeta } from '@ui-kit/lib'
+import { getCampaignsExternalOptions } from './campaigns-external'
+import { getCampaignsMerklOptions } from './campaigns-merkl'
+import type { CampaignPoolRewards, Campaigns } from './types'
+
+/**
+ * Combines multiple campaign records into a single record, merging campaigns by address.
+ *
+ * When the same address appears in multiple campaign sources, all rewards are merged into a single array.
+ * Addresses with no campaigns after filtering are automatically excluded.
+ *
+ * @param campaigns - Array of campaign records to combine (undefined values are filtered out)
+ * @param filter - Optional filter function to apply to individual campaigns before grouping
+ * @returns Combined record with all campaigns merged by address
+ *
+ * @example
+ * ```typescript
+ * // Combine and filter by network
+ * const ethereumCampaigns = combineCampaigns({
+ *   campaigns: [externalCampaigns, merklCampaigns],
+ *   filter: (campaign) => campaign.network === 'ethereum'
+ * })
+ * ```
+ */
+export const combineCampaigns = ({
+  campaigns,
+  filter,
+}: {
+  campaigns: (Campaigns | undefined)[]
+  filter?: (campaign: CampaignPoolRewards) => boolean
+}): Campaigns => {
+  // Get all unique addresses from all campaign sources
+  const allAddresses = new Set(
+    campaigns.filter((record): record is Campaigns => record !== undefined).flatMap((record) => Object.keys(record)),
+  )
+
+  // Combine campaigns by address, applying optional filter
+  return fromEntries(
+    [...allAddresses]
```

Comment:
No reason to copy if you're calling `map`
```suggestion
    allAddresses
```

Reply:
allAddresses is a set which has no `.map`, it needs to be converted to an array first